drafthorse/pdf.py

```diff
@@ -291,3 +258,80 @@ def _facturx_update_metadata_add_attachment(
         )
     metadata_txt_dict = _prepare_pdf_metadata_txt(pdf_metadata)
     pdf_filestream.add_metadata(metadata_txt_dict)
+
+
+def _extract_xml_info(xml_data):
+    """
+    Extract metadata and profile from XML further added to the PDF
+    :param xml_data: XML data
+    :return: Metadata and profile
+    """
+    INVOICE_TYPE_CODE = "380"
+
+    xml_etree = etree.fromstring(xml_data)
+    namespaces = xml_etree.nsmap
+
+    # get metadata
+    date_xpath = xml_etree.xpath(
+        "//rsm:ExchangedDocument/ram:IssueDateTime/udt:DateTimeString",
+        namespaces=namespaces,
+    )
+    date = date_xpath[0].text
+    date_format = date_xpath[0].attrib and date_xpath[0].attrib.get("format") or "102"
+    format_map = {
+        "102": "%Y%m%d",
+        "203": "%Y%m%d%H%M",
+    }
+    date_dt = datetime.strptime(date, format_map.get(date_format, format_map["102"]))
+    number_xpath = xml_etree.xpath(
+        "//rsm:ExchangedDocument/ram:ID", namespaces=namespaces
+    )
+    number = number_xpath[0].text
+    seller_xpath = xml_etree.xpath(
+        "//ram:ApplicableHeaderTradeAgreement/ram:SellerTradeParty/ram:Name",
+        namespaces=namespaces,
+    )
+    seller = seller_xpath[0].text
+
+    doc_type_xpath = xml_etree.xpath(
+        "//rsm:ExchangedDocument/ram:TypeCode", namespaces=namespaces
+    )
+    doc_type = doc_type_xpath[0].text
+    if doc_type != INVOICE_TYPE_CODE:
+        raise Exception(
+            "Invalid doc type! XML value for TypeCode shall be 380 for an invoice."
+        )
+
+    doc_type_name = "Invoice"
+    date_str = datetime.strftime(date_dt, "%Y-%m-%d")
+    pdf_metadata = {
+        "author": seller,
+        "keywords": "{}, {}".format(doc_type_name, "Factur-X"),
+        "title": "{}: {} {}".format(seller, doc_type_name, number),
+        "subject": "{} {} dated {} issued by {}".format(
+            doc_type_name, number, date_str, seller
+        ),
+    }
+
+    # get profile
+    doc_id_xpath = xml_etree.xpath(
+        "//rsm:ExchangedDocumentContext"
+        "/ram:GuidelineSpecifiedDocumentContextParameter"
+        "/ram:ID",
+        namespaces=namespaces,
+    )
+    doc_id = doc_id_xpath[0].text
+
+    profile = doc_id.split(":")[-1]
+    if doc_id.split(":")[-1] in ["basic", "extended"]:
+        profile = doc_id.split(":")[-1]
+    elif doc_id.split(":")[-2] == "en16931":
+        profile = doc_id.split(":")[-2]
+        profile = profile[:2] + " " + profile[2:]
+    else:
+        raise Exception("Invalid XML profile!")
```

Comment:
This is at least missing the "urn:cen.eu:en16931:2017#compliant#urn:xoev-de:kosit:standard:xrechnung_1.2" ID which should be mapped to the "XRECHNUNG" profile.

There might be more profiles out there. I believe we need to keep a way to manually set it instead of only automatically detecting it (but auto-detecting it as a default is nice).